pydm/data_plugins/plugin.py

```diff
@@ -77,85 +79,117 @@ def add_listener(self, channel):
-                except TypeError:
-                    pass
-                try:
-                    self.new_value_signal[bool].disconnect(channel.value_slot)
-                except TypeError:
-                    pass
-
-            if channel.severity_slot is not None:
-                try:
-                    self.new_severity_signal.disconnect(channel.severity_slot)
-                except (KeyError, TypeError):
-                    pass
-
-            if channel.write_access_slot is not None:
-                try:
-                    self.write_access_signal.disconnect(channel.write_access_slot)
-                except (KeyError, TypeError):
-                    pass
-
-            if channel.enum_strings_slot is not None:
-                try:
-                    self.enum_strings_signal.disconnect(channel.enum_strings_slot)
-                except (KeyError, TypeError):
-                    pass
-
-            if channel.unit_slot is not None:
-                try:
-                    self.unit_signal.disconnect(channel.unit_slot)
-                except (KeyError, TypeError):
-                    pass
-
-            if channel.upper_ctrl_limit_slot is not None:
-                try:
-                    self.upper_ctrl_limit_signal.disconnect(channel.upper_ctrl_limit_slot)
-                except (KeyError, TypeError):
-                    pass
-
-            if channel.lower_ctrl_limit_slot is not None:
-                try:
-                    self.lower_ctrl_limit_signal.disconnect(channel.lower_ctrl_limit_slot)
-                except (KeyError, TypeError):
-                    pass
-
-            if channel.prec_slot is not None:
-                try:
-                    self.prec_signal.disconnect(channel.prec_slot)
-                except (KeyError, TypeError):
-                    pass
+    def remove_listener(self, channel, destroying: Optional[bool] = False) -> None:
+        """
+        Removes a listener from this PyDMConnection. If there are no more listeners remaining after
+        removal, then the PyDMConnection will be closed.
+
+        Parameters
+        ----------
+        channel: PyDMChannel
+            The PyDMChannel containing the signals/slots that were being used to listen to the connected address.
+        destroying: bool, optional
+            Should be set to True if this method is being invoked from a flow in which the PyDMWidget using this
+            channel is being destroyed. Since Qt will automatically handle the disconnect of signals/slots when a
+            QObject is destroyed, setting this to True ensures we do not try to do the disconnection a second time.
+            If set to False, any active signals/slots on the channel will be manually disconnected here.
+        """
+        if self._should_disconnect(channel.connection_slot, destroying):
+            try:
+                self.connection_state_signal.disconnect(channel.connection_slot)
+            except TypeError:
```

Comment:
Some of these `except` blocks only catch `TypeError` and some also include `KeyError`. 

I don't know which exception is raised in which scenario, so take this comment/question with a grain of salt: should each of these except blocks be consistently catching both?

Reply:
I couldn't find documentation online about which gets raised when. But doing a quick search through the PyQt5 source only, it seems KeyError is only raised in one place if it can't find the signal at all. I have a feeling that checking for a KeyError was needed prior to the  `destorying` kwarg being introduced, and that the correct action would be to not catch it at all anymore from any of these disconnects.

However I think I will err on the side of caution and leave it as is for the next release. This block of code will most likely change quite a bit with potential PyDM 2.0 work, and I will add verifying exactly which error is needed to check for to the task list for the plugin changes.